pageserver/src/tenant/timeline.rs

```diff
@@ -4471,14 +4473,30 @@ impl Timeline {
             let min_distance = self.get_image_layer_creation_check_threshold() as u64
                 * self.get_checkpoint_distance();
 
-            // Skip the expensive delta layer counting if this timeline has not ingested sufficient
-            // WAL since the last check.
-            distance.0 >= min_distance
-        };
+            let distance_based_decision = distance.0 >= min_distance;
 
-        if check_for_image_layers {
-            self.last_image_layer_creation_check_at.store(lsn);
-        }
+            let mut last_check_instant =
+                self.last_image_layer_creation_check_instant.lock().unwrap();
+            let time_based_decision = match *last_check_instant {
```

Comment:
Thinking of how this changes edge cases: up til now, a timeline whose LSN had never risen above the check threshold could never generate image layers.  I think that works out to 256MB with default settings.

In the pathological case where someone has a 32MB database that they write like 1kb to every 10 minutes, we can generate image layers based on the time threshold.

Maybe we should explicitly avoid generating image layers in this pathological case where a tiny database is doing tiny writes, on the basis that this kind of database definitely fits in RAM and doesn't motivate us to optimize read performance by generating image layers.

Reply:
> I think that works out to 256MB with default settings

Should be 512MiB

> In the pathological case where someone has a 32MB database that they write like 1kb to every 10 minutes, we can generate image layers based on the time threshold.

Well we still need to find three stacked deltas in order to create an image layer. In the pathological scenario you are describing, we'd create an image layer every 30 minutes.

> Maybe we should explicitly avoid generating image layers in this pathological case where a tiny database is doing tiny writes, on the basis that this kind of database definitely fits in RAM and doesn't motivate us to optimize read performance by generating image layers.

So you are saying that stacking loads of tiny delta layers without image layers isn't actually problematic? Doesn't it inhibit GC (not a big deal - the db is small) and make branching problematic (translating timestamps to lsn)?

Reply:
> Well we still need to find three stacked deltas in order to create an image layer. In the pathological scenario you are describing, we'd create an image layer every 30 minutes.

I thought it was more like this:
- L0 layer every 10 minutes
- L1 layers cover full keyspace, generated every 10 L0 layers, so every 100 minutes
- three L1 layers trigger image layer generation, so we'd generate image layers every ~5 hours.

>So you are saying that stacking loads of tiny delta layers without image layers isn't actually problematic?

It's definitely problematic in general, but I'm saying maybe we should special case tiny tenants with tiny layers, because (some hand-waving here):
- Pageserver performance for tiny tenants is relatively unimportant, as they will rarely spill out of their compute's RAM
- Tiny tenants are very unlikely to be used in serious applications: users simply care less about performance
- Doing time->LSN lookup is not going to be super expensive because under low write rates there won't be so many distinct LSNs to seek through, and at each LSN there won't be that many CLog entries that we need to inspect.
- A read that spans 50 tiny delta layers isn't that expensive unless all those tiny layers contain deltas, but:
   - the tiny indices are cheap to traverse, it isn't like having to traverse 50 full-sized layers
   - postgres checkpoints generating image values are fairly likely to have happened during this time period.

But my underlying take would be that if we were previously not generating image layers for such tenants, and nobody was complaining, we should probably avoid generating lots of image layers.  If we can get confident that having lots more deltas (up to some bound, like generating image layers every few days) for tiny tenants isn't harmful, then the right fix might just be to make our metrics/alerting conditional on the size of the tenant.

Reply:
> I thought it was more like this:
> * L0 layer every 10 minutes
> * L1 layers cover full keyspace, generated every 10 L0 layers, so every 100 minutes
> * three L1 layers trigger image layer generation, so we'd generate image layers every ~5 hours

Yes, that's right. Before this patch, we did not generate image layers for slow ingest tenants. After this patch, we would generate one every 5 hours. Worth noting that 4 months or so ago, before introducing the WAL based check we did generate image layers.

> It's definitely problematic in general, but I'm saying maybe we should special case tiny tenants with tiny layers, because (some hand-waving here):

I don't disagree with your points, but there's the counter-argument that we are doing loads of disk io which also has implications.

> But my underlying take would be that if we were previously not generating image layers for such tenants, and nobody was complaining, we should probably avoid generating lots of image layers. If we can get confident that having lots more deltas (up to some bound, like generating image layers every few days) for tiny tenants isn't harmful, then the right fix might just be to make our metrics/alerting conditional on the size of the tenant.

Understood. Why only tiny tenants though? Many of the same arguments apply to large tenants that have low ingest tput. If this insight is right, we could just make this time based check threshold 24h or 48h.

Reply:
>Why only tiny tenants though? Many of the same arguments apply to large tenants that have low ingest tput. If this insight is right, we could just make this time based check threshold 24h or 48h.

The main reasons I would avoid this for large tenants are that  large tenants are much more likely to care about getpage performance because:
- their workloads might not fit in RAM
- they're much more likely to be someone's real production database

Reply:
Okay, how about this: for timelines with a logical size greater than 2GiB, we force a check every 10 minutes (checkpoint timeout) and for anything smaller we do it every 48h. Numbers can be debated.

Reply:
Sounds fine to me.

Reply:
Done in https://github.com/neondatabase/neon/pull/8247/commits/2c1884d00857f28083f57291675bab71f391502d

Reply:
This is certainly easier to implement than the "L0 -> L0" compaction.

Are the larger-than-threshold timelines that switch modes from "producing a lot of wal" to "producing wal like testodrome project" now a problem which we tried to avoid with the original Lsn distance based threshold for image layer creation?

Reply:
> Are the larger-than-threshold timelines that switch modes from "producing a lot of wal" to "producing wal like testodrome project" now a problem which we tried to avoid with the original Lsn distance based threshold for image layer creation?

No, it wasn't something that was considered at the time.

For such "switchy" tenants we would indeed have greater storage amplification (i.e. produce image layers for small amounts of ingested WAL). John had some points above on why it makes sense to swallow up the cost in these cases:

> The main reasons I would avoid this for large tenants are that large tenants are much more likely to care about getpage performance because:
> * their workloads might not fit in RAM
> *  they're much more likely to be someone's real production database